### PR TITLE
Fix cusparseSpMV warnings for CUDA-11.2

### LINKED ISSUE #2164: Warnings with cuda 11.2.2
Hello, I observe the following warnings when I upgrade to cuda 11.2

linalg/sparsemat.cpp(679): warning #1786: constant "CUSPARSE_CSRMV_ALG1" (declared at line 6314 of "/usr/local/cuda-11.2/bin/../targets/x86_64-linux/include/cusparse.h") was declared deprecated ("please use CUSPARSE_SPMV_CSR_ALG1 instead")
  cusparseSpMV_bufferSize(handle, CUSPARSE_OPERATION_NON_TRANSPOSE, &alpha, matA_descr, vecX_descr, &beta, vecY_descr, CUDA_R_64F, CUSPARSE_CSRMV_ALG1, &newBufferSize); 
                                                                                                                                   ^

linalg/sparsemat.cpp(697): warning #1786: constant "CUSPARSE_CSRMV_ALG1" (declared at line 6314 of "/usr/local/cuda-11.2/bin/../targets/x86_64-linux/include/cusparse.h") was declared deprecated ("please use CUSPARSE_SPMV_CSR_ALG1 instead")
  cusparseSpMV(handle, CUSPARSE_OPERATION_NON_TRANSPOSE, &alpha, matA_descr, vecX_descr, &beta, vecY_descr, CUDA_R_64F, CUSPARSE_CSRMV_ALG1, dBuffer); 
                                                                                                                        ^

OS: Ubuntu 18.04
CXX: mpiicpc
CUDA: cuda_11.2.2_460.32.03_linux.run

### PR DESCRIPTION
**Pull request by adam-sim-dev**
Resolves #2164.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2200](https://github.com/mfem/mfem/pull/2200) | @adam-sim-dev | @tzanio | @YohannDudouit + @artv3 | 06/02/21 | 06/03/21 | 06/10/21 | |

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by tzanio** [comment]
LGTM

**Review by artv3 — APPROVED** [review]

**Comment by wcdawn** [comment]
Compiles on my machine using CUDA 11.3. Thanks!

**Review by wcdawn — APPROVED** [review]

**Review by YohannDudouit — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
We got this error on Lassen in tonight's test:

```
linalg/sparsemat.cpp(694): error: identifier "CUSPARSE_SPMV_CSR_ALG1" is undefined
1 error detected in the compilation of "linalg/sparsemat.cpp".
```

Any suggestions how to fix?

**Comment by adrienbernede** [comment]
Note: The error is while building with `cuda@11.1.1`, and `-arch=sm_70` (using spack and xl-16.1.1.8 compiler).

**Comment by YohannDudouit** [comment]
> We got this error on Lassen in tonight's test:
> 
> ```
> linalg/sparsemat.cpp(694): error: identifier "CUSPARSE_SPMV_CSR_ALG1" is undefined
> 1 error detected in the compilation of "linalg/sparsemat.cpp".
> ```
> 
> Any suggestions how to fix?

I'm actually surprised by this, I tried it with both an old cuda and a recent cuda on lassen. Could it be that the `CUDA_VERSION` check is erroneous, `#if CUDA_VERSION > 10020 || CUDA_VERSION == 10020` should rather be something like `#if CUDA_VERSION > 11020 || CUDA_VERSION == 11020`?

File: linalg/sparsemat.cpp [hunk 1]
```diff
@@ -687,7 +687,12 @@ void SparseMatrix::AddMult(const Vector &x, Vector &y, const double a) const
       // Allocate kernel space. Buffer is shared between different sparsemats
       size_t newBufferSize = 0;
 
-#if CUDA_VERSION > 10010 || CUDA_VERSION == 10010
+#if CUDA_VERSION > 10020 || CUDA_VERSION == 10020
```
Comment (YohannDudouit) [inline linalg/sparsemat.cpp:690]:
```suggestion
#if CUDA_VERSION > 11020 || CUDA_VERSION == 11020
```

File: linalg/sparsemat.cpp [hunk 2]
```diff
@@ -702,7 +707,15 @@ void SparseMatrix::AddMult(const Vector &x, Vector &y, const double a) const
          CuMemAlloc(&dBuffer, bufferSize);
       }
 
-#if CUDA_VERSION > 10010 || CUDA_VERSION == 10010
+#if CUDA_VERSION > 10020 || CUDA_VERSION == 10020
```
Comment (YohannDudouit) [inline linalg/sparsemat.cpp:710]:
```suggestion
#if CUDA_VERSION > 11020 || CUDA_VERSION == 11020
```

**Review by YohannDudouit — CHANGES_REQUESTED** [review]
This seems to fix the issue on Lassen.

File: linalg/sparsemat.cpp (hunk 1 above)
Comment (adrienbernede) [reply linalg/sparsemat.cpp:690]: A laborious search in the cuda documentation _almost_ proves @YohannDudouit right.
Per documentation, `CUSPARSE_CSRMV_ALG1` was deprecated in `11.2.1`, when `CUSPARSE_SPMV_CSR_ALG1` was introduced. **It seems this should not work with 11.2.0, but only with 11.2.1 and higher**.
See:
https://docs.nvidia.com/cuda/archive/11.2.0/cusparse/index.html#cusparse-generic-api-reference
https://docs.nvidia.com/cuda/archive/11.2.1/cusparse/index.html#cusparse-generic-api-reference

File: linalg/sparsemat.cpp (hunk 2 above)
Comment (adrienbernede) [reply linalg/sparsemat.cpp:710]: A laborious search in the cuda documentation _almost_ proves @YohannDudouit right.
Per documentation, `CUSPARSE_CSRMV_ALG1` was deprecated in `11.2.1`, when `CUSPARSE_SPMV_CSR_ALG1` was introduced. **It seems this should not work with 11.2.0, but only with 11.2.1 and higher**.
See:
https://docs.nvidia.com/cuda/archive/11.2.0/cusparse/index.html#cusparse-generic-api-reference
https://docs.nvidia.com/cuda/archive/11.2.1/cusparse/index.html#cusparse-generic-api-reference

File: linalg/sparsemat.cpp (hunk 1 above)
Comment (adrienbernede) [reply linalg/sparsemat.cpp:690]: @YohannDudouit did you really run with 11.2.0, and not with 11.2.1 ?
It is possible that `CUSPARSE_SPMV_CSR_ALG1` was in the code before it was documented.

Comment (tomstitt) [reply linalg/sparsemat.cpp:690]: I ran into this issue too and found that `11.2.0` defines `CUSPARSE_SPMV_CSR_ALG1`.

On LC see `/usr/tce/packages/cuda/cuda-11.2.0/include/cusparse.h`

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

File: linalg/sparsemat.cpp [hunk 3]
```diff
@@ -687,7 +687,12 @@ void SparseMatrix::AddMult(const Vector &x, Vector &y, const double a) const
       // Allocate kernel space. Buffer is shared between different sparsemats
       size_t newBufferSize = 0;
 
-#if CUDA_VERSION > 10010 || CUDA_VERSION == 10010
+#if CUDA_VERSION > 11020 || CUDA_VERSION == 11020
```
Comment (v-dobrev) [inline linalg/sparsemat.cpp:690]: Is there a reason to not use `CUDA_VERSION >= 11020` here and in other places?

Comment (tzanio) [reply linalg/sparsemat.cpp:690]: @adam-sim-dev  we will replace all these instances with `>=` as @v-dobrev suggested. Please speak up if you have any concerns.

Comment (tzanio) [reply linalg/sparsemat.cpp]: ☝️  implemented in  https://github.com/mfem/mfem/pull/2200/commits/914c334a8b6342c7b3abf298a91552198bd3ce48

**Comment by tzanio** [comment]
Re-merged in `next` for testing ...

**Review by tzanio — APPROVED** [review]
Thanks for this fix @adam-sim-dev !

**Pull request merged**